contracts/GSDINFT.sol

```diff
@@ -0,0 +1,200 @@
+  /// @param _faceValue Amount of currency borrower must pay to cover the GSDI.
+  /// @param _price Amount of currency the lender must pay the borrower to mint the GSDI.
+  /// @param _wallet Wallet containing collateral backing the GSDI.
+  /// @param _currency Token currency for the price and face value.
+  /// @param _borrower Address which will become the wallet executor after the GSDI is covered.
+  function propose(
+    uint256 _maturity,
+    uint256 _faceValue,
+    uint256 _price,
+    IGSDIWallet _wallet,
+    address _currency,
+    address _borrower
+  ) external override returns (uint256 id_) {
+    uint256 tokenId = _tokenIdTracker.current(); 
+    id_ = abi.encodePacked(tokenId >> 96, chainId << 160).toUint256(0);
+
+    metadata[id_] = NFTMetadata(_maturity, _faceValue, _price, _wallet, _currency, _borrower, true);
+
+    // Locks the IGSDIWallet by setting IGSDINFT as the wallet's executor.
+    _wallet.setExecutor(address(this));
+
+    _mint(_borrower, id_);
+    _tokenIdTracker.increment();
+
+    emit Propose(id_, tokenId, _borrower, address(this));
+  }
+
+  ///@notice Cancels the GSDI proposal. Sender must be borrower. GSDI must currently be in proposal. GSDI must be on the current chain. Burns the GSDI.
+  /// @param _id GSDI to cancel.
+  function cancel(uint256 _id) external mustBeInProposal(_id) override {
+    burnProposal(_id);
+    
+    emit Cancel(_id);
+  }
+
+  /// @notice Sends the GSDI to sender. The IGSDIWallet must be in proposal. GSDI must be on the current chain. Transfers price in currency from sender to borrower. Removes GSDI from proposal.
+  /// @notice If governance has enabled the 0.3% fee, the sender must also transfer the fee to the governance's treasury.
+  /// @dev Sender must approve the contract to transfer price of currency before call.
+  /// @param _id GSDI to purchase.
+  function purchase(uint256 _id) external mustBeInProposal(_id) override {
+    uint256 price = metadata[_id].price;
+    
+    IERC20 token = IERC20(metadata[_id].currency);
+    if (token.allowance(msg.sender, address(this)) < price) {
+      token.approve(address(this), price);
+    }
+
+    bytes memory transferPayload = abi.encodeWithSignature("transferFrom(address,address,unint256)", msg.sender, metadata[_id].borrower, price);
+    IGSDIWallet(metadata[_id].wallet).execute(metadata[_id].borrower, 1 ether, transferPayload);
+    metadata[_id].isInProposal = false;
+
+    emit Purchase(_id, msg.sender, price, metadata[_id].borrower);
+  }
+
+  /// @notice Sets the borrower for IGSDIWallet to executor. The IGSDIWallet must not be in proposal. GSDI must be on the current chain. Sender transfers face value in currency to current GSDI holder. Burns the GSDI.
+  /// @dev Sender must approve the contract to transfer face value of currency before call.
+  /// @param _id GSDI to cover.
+  function cover(uint256 _id) external override mustNotBeInProposal(_id) {
```

Comment:
Must revert if IGSDIWallet's chainID is different from the GSDINFT contract's chainID.